File: python/prob_003.py
# prob_003.py

# Problem 3
#
# The prime factors of 13195 are 5, 7, 13 and 29.
#
# What is the largest prime factor of the number 600851475143 ?
#

from Primes import Primes 

def solve(number, primes):
    # loop over primes, check which are divisors
    # search up to number in case number is prime
    primeFactors = []
    i = 0
    while i < len(primes) and primes[i] <= number:
        if number % primes[i] == 0:
            primeFactors.append(primes[i])
        i += 1
    return primeFactors

max_val = 10**7
P = Primes()
#P.calcPrimesBasic(max_val)
P.calcPrimesAdvanced(max_val)
primes = P.getPrimes()
print("number of primes less than {0}: {1}".format(max_val, len(primes)))

#x = solve(13195, primes)
#print("prime factors: {0}".format(x))
#print("largest prime factor: {0}".format(max(x)))

x = solve(600851475143, primes)
print("prime factors: {0}".format(x))
print("largest prime factor: {0}".format(max(x)))

#numbers = [
#    101,
#    10101,
#    1010101,
#    101010101,
#    10101010101,
#]
#for n in numbers:
#    print("{0}: {1}".format(n, solve(n, primes)))

File: rust/prob_003.rs
// prob_003.rs

// Problem 3
// 
// Largest Prime Factor
// 
// The prime factors of 13195 are 5, 7, 13 and 29.
// 
// What is the largest prime factor of the number 600851475143 ?
// 

// import standard library types
use std::time::Instant;

// Solve the problem
fn solve(max_n: i32) -> i32 {
    println!("Running solve()... go go go!");
    println!("max_n: {max_n}");
    
    // Since we plan to change result, we need to make it mutable using "mut".
    let mut result: i32 = 0;
    
    println!("Completed solve()!");
    result
}

// This is the main function.
fn main() {
    let start_time  = Instant::now();
    let max_n: i32  = 1_000_000;
    let answer: i32 = solve(max_n);
    let end_time    = Instant::now();
    
    let run_time = end_time.duration_since(start_time);

    println!("answer: {answer}");
    println!("run time: {:.2?}", run_time);
}



